pkg/daemon/daemon.go

```diff
@@ -747,21 +766,116 @@ func (dn *Daemon) annotateNode(node, value string) error {
 	return nil
 }
 
+func (dn *Daemon) getNodeMachinePool() error {
+	mcpList, err := dn.mcClient.MachineconfigurationV1().MachineConfigPools().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		glog.Errorf("getNodeMachinePool(): failed to list Machine Config Pools: %v", err)
+		return err
+	}
+	var mcp mcfgv1.MachineConfigPool
+	for _, mcp = range mcpList.Items {
+		selector, err := metav1.LabelSelectorAsSelector(mcp.Spec.NodeSelector)
+		if err != nil {
+			glog.Errorf("getNodeMachinePool(): Machine Config Pool %s invalid label selector: %v", mcp.GetName(), err)
+			return err
+		}
+
+		if selector.Matches(labels.Set(dn.node.Labels)) {
+			dn.mcpName = mcp.GetName()
+			glog.Infof("getNodeMachinePool(): find node in MCP %s", dn.mcpName)
+			return nil
+		}
+	}
+	return fmt.Errorf("getNodeMachinePool(): fail to find the MCP of the node")
+}
+
 func (dn *Daemon) drainNode(name string) error {
 	glog.Info("drainNode(): Update prepared")
 	var err error
+
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
 	// wait a random time to avoid all the nodes drain at the same time
-	wait.PollUntil(time.Duration(rand.Intn(15)+1)*time.Second, func() (bool, error) {
+	time.Sleep(wait.Jitter(3*time.Second, 3))
+	wait.JitterUntil(func() {
 		if !dn.drainable {
-			glog.Info("drainNode(): other node is draining, waiting...")
+			glog.V(2).Info("drainNode(): other node is draining")
+			return
+		}
+		glog.V(2).Info("drainNode(): no other node is draining")
+		err = dn.annotateNode(dn.name, annoDraining)
+		if err != nil {
+			glog.Errorf("drainNode(): Failed to annotate node: %v", err)
+			return
+		}
+		cancel()
+	}, 3*time.Second, 3, true, ctx.Done())
+
+	if utils.ClusterType == utils.ClusterTypeOpenshift {
+		mcpInformerFactory := mcfginformers.NewSharedInformerFactory(dn.mcClient,
+			time.Second*30,
+		)
+		mcpInformer := mcpInformerFactory.Machineconfiguration().V1().MachineConfigPools().Informer()
+
+		ctx, cancel := context.WithCancel(context.TODO())
+		defer cancel()
+		paused := false
+		mcpEventHandler := func(obj interface{}) {
+			mcp := obj.(*mcfgv1.MachineConfigPool)
+			if mcp.GetName() != dn.mcpName {
+				return
+			}
+			// Always get the latest object
+			newMcp, err := dn.mcClient.MachineconfigurationV1().MachineConfigPools().Get(ctx, dn.mcpName, metav1.GetOptions{})
+			if err != nil {
+				glog.V(2).Infof("drainNode(): fail to get MCP %s: %v", dn.mcpName, err)
+				return
+			}
+			if mcfgv1.IsMachineConfigPoolConditionFalse(newMcp.Status.Conditions, mcfgv1.MachineConfigPoolDegraded) &&
+				mcfgv1.IsMachineConfigPoolConditionTrue(newMcp.Status.Conditions, mcfgv1.MachineConfigPoolUpdated) &&
+				mcfgv1.IsMachineConfigPoolConditionFalse(newMcp.Status.Conditions, mcfgv1.MachineConfigPoolUpdating) {
+				glog.V(2).Infof("drainNode(): MCP %s is ready", dn.mcpName)
+				if paused {
+					glog.V(2).Info("drainNode(): stop MCP informer", dn.mcpName)
+					cancel()
+					return
+				}
+				if newMcp.Spec.Paused {
```

Comment:
will this be the case if the sriov operator drains the node and then reboot it to change for example the mlx max number of VFs?

Reply:
After reboot, the config daemon will not drain the node, so we will not hit this logic.

Reply:
Hi @pliurh did you try this on an intel nic?

I am waiting to get an env but I think in general for intel

1. add the iommu in the kargs
2. drainNode code https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/f552cd8eb8990787a24a0504f633405adfd2af76/pkg/daemon/daemon.go#L535
3. reboot https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/f552cd8eb8990787a24a0504f633405adfd2af76/pkg/daemon/daemon.go#L551
4. when the node is up again the default plugin will request to drain again https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/f552cd8eb8990787a24a0504f633405adfd2af76/pkg/plugins/generic/generic_plugin.go#L86
5. to configure the numOfVfs https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/f552cd8eb8990787a24a0504f633405adfd2af76/pkg/plugins/generic/generic_plugin.go#L193

it will stuck because the mcp is already pause do I miss something?

Reply:
In step5, It will not require to drain a node, as the `ifaceStatus.NumVfs` is always 0 when a node boots up.

However, there is still a case that can trigger the problem, when a non-default MTU is specified. I'll fix that.